txnsync/txngroups.go

```diff
@@ -58,3 +227,522 @@ func decodeTransactionGroups(bytes []byte) (txnGroups []transactions.SignedTxGro
 	}
 	return txnGroups, nil
 }
+
+func encodeTransactionGroups(inTxnGroups []transactions.SignedTxGroup) []byte {
+	txnCount := 0
+	for _, txGroup := range inTxnGroups {
+		txnCount += len(txGroup.Transactions)
+	}
+	stub := txGroupsEncodingStub{
+		TotalTransactionsCount: uint64(txnCount),
+		TransactionGroupCount:  uint64(len(inTxnGroups)),
+		TransactionGroupSizes:  make([]byte, 0, len(inTxnGroups)), // compress this later to use 2 per byte
+		SignedTxns:             make([]transactions.SignedTxn, 0, txnCount),
+	}
+	for _, txGroup := range inTxnGroups {
+		if len(txGroup.Transactions) > 1 {
+			for _, txn := range txGroup.Transactions {
+				stub.SignedTxns = append(stub.SignedTxns, txn)
+			}
+			stub.TransactionGroupSizes = append(stub.TransactionGroupSizes, byte(len(txGroup.Transactions)-1))
+		}
+	}
+	for _, txGroup := range inTxnGroups {
+		if len(txGroup.Transactions) == 1 {
+			for _, txn := range txGroup.Transactions {
+				stub.SignedTxns = append(stub.SignedTxns, txn)
+			}
+		}
+	}
+	for i := range stub.SignedTxns {
+		stub.SignedTxns[i].Txn.Group = crypto.Digest{}
+	}
+
+	deconstructSignedTransactions(&stub)
+	return stub.MarshalMsg([]byte{})
+}
+
+func decodeTransactionGroups(bytes []byte) (txnGroups []transactions.SignedTxGroup, err error) {
+	if len(bytes) == 0 {
+		return nil, nil
+	}
+	var stub txGroupsEncodingStub
+	_, err = stub.UnmarshalMsg(bytes)
+	if err != nil {
+		return nil, err
+	}
+
+	reconstructSignedTransactions(&stub)
+
+	txnGroups = make([]transactions.SignedTxGroup, stub.TransactionGroupCount)
+	for index, i := 0, 0; index < int(stub.TotalTransactionsCount); i++ {
+		size := 1
+		if i < len(stub.TransactionGroupSizes) {
+			size = int(stub.TransactionGroupSizes[i]) + 1
+		}
+		txnGroups[i].Transactions = stub.SignedTxns[index : index+size]
+		index += size
+	}
+
+	addGroupHashes(txnGroups)
+
+	return txnGroups, nil
+}
+
+func addGroupHashes(txnGroups []transactions.SignedTxGroup) {
+	for _, txns := range txnGroups {
+		if len(txns.Transactions) > 1 {
```

Comment:
this is abit more tricky : 
A transaction group of size 1 is allowed to have a transaction group field. so, if you were to add a bitmask specifically for that case, it would be great.